blackjax/mcmc/elliptical_slice.py

```diff
@@ -0,0 +1,166 @@
+
+    cov_matrix_sqrt = jax.lax.linalg.cholesky(cov_matrix)
+
+    def momentum_generator(rng_key, position):
+        p, unravel_fn = jax.flatten_util.ravel_pytree(position)
+        momentum = mean + jnp.dot(
+            cov_matrix_sqrt, jax.random.normal(rng_key, shape=p.shape)
+        )
+        return unravel_fn(momentum)
+
+    def one_step(
+        rng_key: PRNGKey,
+        state: EllipSliceState,
+        loglikelihood_fn: Callable,
+    ) -> Tuple[EllipSliceState, EllipSliceInfo]:
+        proposal_generator = elliptical_proposal(
+            loglikelihood_fn, momentum_generator, mean
+        )
+        return proposal_generator(rng_key, state)
+
+    return one_step
+
+
+def elliptical_proposal(
+    loglikelihood_fn: Callable,
+    momentum_generator: Callable,
+    mean: Array,
+) -> Callable:
+    """Build an Ellitpical slice sampling kernel.
+
+    The algorithm samples a latent parameter, traces an ellipse connecting the
+    initial position and the latent parameter and does slice sampling on this
+    ellipse to output a new sample from the posterior distribution.
+
+    Parameters
+    ----------
+    loglikelihood_fn
+        A function that returns the log-likelihood at a given position.
+    momentum_generator
+        A function that generates a new latent momentum variable.
+
+    Returns
+    -------
+    A kernel that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+
+    """
+
+    def generate(
+        rng_key: PRNGKey, state: EllipSliceState
+    ) -> Tuple[EllipSliceState, EllipSliceInfo]:
+        position, loglikelihood = state
+        kmomentum, kunif, ktheta = jax.random.split(rng_key, 3)
+        # step 1: sample momentum
+        momentum = momentum_generator(kmomentum, position)
+        # #step 2: get slice (y)
+        logy = loglikelihood + jnp.log(jax.random.uniform(kunif))
+        # step 3: get theta (ellipsis move), set inital interval
+        theta = 2 * jnp.pi * jax.random.uniform(ktheta)
+        theta_min = theta - 2 * jnp.pi
+        theta_max = theta
+        # step 4: proposal
+        p, m = ellipsis(position, momentum, theta, mean)
+        # step 5: acceptance
+        loglikelihood = loglikelihood_fn(p)
+
+        def while_fun(vals):
+            rng, _, subiter, theta, theta_min, theta_max, *_ = vals
+            rng, thetak = jax.random.split(rng)
+            theta = jax.random.uniform(thetak, minval=theta_min, maxval=theta_max)
+            p, m = ellipsis(position, momentum, theta, mean)
+            loglikelihood = loglikelihood_fn(p)
+            theta_min = jnp.where(theta < 0, theta, theta_min)
+            theta_max = jnp.where(theta > 0, theta, theta_max)
+            subiter += 1
+            return rng, loglikelihood, subiter, theta, theta_min, theta_max, p, m
+
+        _, loglikelihood, subiter, theta, *_, position, momentum = jax.lax.while_loop(
```

Comment:
Is this guaranteed to stop or are there circumstances where we can get stuck in an infinite loop?

Reply:
As the bracket `[theta_min, theta_max]` shrinks, the proposal gets closer to the original position with likelihood `log y < loglikelihood_fn(position)`. It is guaranteed to stop in a finite number of iterations as long as the log likelihood is continuous wrt the parameter we are sampling (since the parameter is a continuous random variable with Gaussian prior I'm pretty sure its guaranteed to stop in finite iterations always).

Reply:
I see. Could you add this in the docstring of the body function of the while loop?